packages/sanity/src/core/store/_legacy/document/getPairListener.ts

```diff
@@ -79,18 +87,55 @@ export function getPairListener(
           effectFormat: 'mendoza',
           tag: options.tag || 'document.pair-listener',
         },
-      ) as Observable<WelcomeEvent | MutationEvent | ReconnectEvent>,
-  ).pipe(
-    concatMap((event) =>
-      event.type === 'welcome'
+      )
+      .pipe(
+        shareReplayLatest({
+          predicate: (event) => event.type === 'welcome' || event.type === 'reconnect',
+          resetOnRefCountZero: () => timer(LISTENER_RESET_DELAY),
+        }),
+      ),
+  ) as Observable<WelcomeEvent | MutationEvent | ReconnectEvent>
+
+  return defer(() => sharedEvents).pipe(
+    concatMap((event) => {
+      return event.type === 'welcome'
         ? fetchInitialDocumentSnapshots().pipe(
-            concatMap((snapshots) => [
-              createSnapshotEvent(draftId, snapshots.draft),
-              createSnapshotEvent(publishedId, snapshots.published),
-            ]),
+            concatMap((snapshots) => {
+              return merge(
+                concat(
+                  of(createSnapshotEvent(draftId, snapshots.draft)),
+                  sharedEvents.pipe(
+                    filter((event) => event.type !== 'welcome'),
+                    filter(
+                      (sharedListenerEvent) =>
+                        sharedListenerEvent.type !== 'mutation' ||
+                        sharedListenerEvent.documentId === draftId,
+                    ),
+                  ),
+                ).pipe(sequentializeListenerEvents()),
+                concat(
+                  of(createSnapshotEvent(publishedId, snapshots.published)),
+                  sharedEvents.pipe(
+                    filter((event) => event.type !== 'welcome'),
+                    filter(
+                      (sharedListenerEvent) =>
+                        sharedListenerEvent.type !== 'mutation' ||
+                        sharedListenerEvent.documentId === publishedId,
+                    ),
+                  ),
+                ).pipe(sequentializeListenerEvents()),
+              )
+            }),
           )
-        : observableOf(event),
-    ),
+        : EMPTY
+    }),
+    catchError((err, caught$) => {
+      if (err instanceof OutOfSyncError) {
+        console.error(err)
```

Comment:
the callback idea you mentioned to be would be nice to have here so we can add telemetry for when this happens